### PR TITLE
[feature](external catalog)Add partition grammar for external catalog  to create table

### PR DESCRIPTION
**Pull request by wuwenchi**
## Proposed changes

The `PARTITION BY` syntax used by external catalogs has been added. 
You can specify a column directly, or a partition function as a partition condition.
Like:
 `PARTITION BY LIST(col1, col2, func(param), func(param1, param2), func(param1, param2, param3))`

NOTICE:
This PR change the grammar of `AUTO PARTITION`
From 
```
AUTO PARTITION BY RANGE date_trunc(`TIME_STAMP`, 'month')
```
To
```
AUTO PARTITION BY RANGE (date_trunc(`TIME_STAMP`, 'month'))
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 [hunk 1]
```diff
@@ -717,6 +722,21 @@ primaryExpression
...
 
+partitionFunctionCallExpression
+    : LEFT_PAREN partitionFunctionList RIGHT_PAREN
+    ;
+
+partitionFunctionList
+    : functions+=partitionFunction (COMMA functions+=partitionFunction)*
+    ;
+
+partitionFunction
```
Comment (morrySnow) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:733]: it is better use `functionCallExpression` directly, and in LogicalPlanBuilder check return expression is `UnboundFunction`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 2]
```diff
@@ -2512,6 +2492,75 @@ public List<ColumnDefinition> visitColumnDefs(ColumnDefsContext ctx) {
...
+    public List<Expression> visitPartitionFunctionList(DorisParser.PartitionFunctionListContext ctx) {
+        return ctx.functions.stream()
+            .map(this::visitPartitionFunction)
+            .collect(ImmutableList.toImmutableList());
+    }
+
+    @Override
+    public Expression visitPartitionFunction(DorisParser.PartitionFunctionContext ctx) {
+        String name = ctx.identity.getText();
+        List<Expression> params = visit(ctx.expression(), Expression.class);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:2560]: do u test this syntax? i think u want get is ctx.arguments()

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 [hunk 3]
```diff
@@ -717,6 +722,21 @@ primaryExpression
...
     ;
 
+partitionFunctionCallExpression
+    : LEFT_PAREN partitionFunctionList RIGHT_PAREN
+    ;
+
+partitionFunctionList
+    : functions+=partitionFunction (COMMA functions+=partitionFunction)*
+    ;
+
```
Comment (morrySnow) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:732]: why not merge this two statement into one?
```
LEFT_PAREN functions+=partitionFunction (COMMA functions+=partitionFunction)* RIGHT_PAREN
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 4]
```diff
@@ -2470,11 +2450,11 @@ public LogicalPlan visitCreateTable(CreateTableContext ctx) {
                     keysType,
                     ctx.keys != null ? visitIdentifierList(ctx.keys) : ImmutableList.of(),
                     "",
-                    isAutoPartition,
-                    autoPartitionExpr.build(),
-                    partitionType,
-                    ctx.partitionKeys != null ? visitIdentifierList(ctx.partitionKeys) : null,
-                    ctx.partitions != null ? visitPartitionsDef(ctx.partitions) : null,
+                    partitionInfo.isAutoPartition(),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:2453]: if u aready merge all partition info into one class, u'd better update `CreateTableInfo`'s ctor to accept `PartitionTableInfo` as parameter

**Comment by wuwenchi** [comment]
run buildall

**Comment by wuwenchi** [comment]
run buildall

**Comment by wuwenchi** [comment]
run buildall

**Comment by wuwenchi** [comment]
run buildall

**Comment by wuwenchi** [comment]
run buildall

**Review by morningman — DISMISSED** [review]
LGTM

**Comment by wuwenchi** [comment]
run buildall

**Comment by morningman** [comment]
run buildall

**Comment by morningman** [comment]
run buildall

**Comment by wuwenchi** [comment]
run buildall

**Comment by wuwenchi** [comment]
run p0

**Comment by wuwenchi** [comment]
run p0

**Review by morningman — APPROVED** [review]
LGTM

**Review by kaka11chen — APPROVED** [review]
LGTM.

**Pull request merged**